Lib/pathlib.py

```diff
@@ -207,36 +206,6 @@ def _select_from(self, parent_path, is_dir, exists, scandir, normcase):
 # Public API
 #
 
-class _PathParents(Sequence):
-    """This object provides sequence-like access to the logical ancestors
-    of a path.  Don't try to construct it yourself."""
-    __slots__ = ('_pathcls', '_drv', '_root', '_tail')
-
-    def __init__(self, path):
-        # We don't store the instance to avoid reference cycles
-        self._pathcls = type(path)
-        self._drv = path.drive
-        self._root = path.root
-        self._tail = path._tail
-
-    def __len__(self):
-        return len(self._tail)
-
-    def __getitem__(self, idx):
-        if isinstance(idx, slice):
-            return tuple(self[i] for i in range(*idx.indices(len(self))))
-
-        if idx >= len(self) or idx < -len(self):
-            raise IndexError(idx)
-        if idx < 0:
-            idx += len(self)
-        return self._pathcls._from_parsed_parts(self._drv, self._root,
-                                                self._tail[:-idx - 1])
-
-    def __repr__(self):
-        return "<{}.parents>".format(self._pathcls.__name__)
```

Comment:
The current implementation of `_PathParents` goes to great lengths to only create path objects "lazily" on request. That means that replacing `_PathParents` with a simple tuple creates quite a bad performance regression for this (unrealistic, artificial) benchmark.

On `main`:

```
>python -m timeit -s "from pathlib import Path; p = Path(*['foo' for _ in range(1_000)])" "p.parents[100]"
Running Debug|x64 interpreter...
2000 loops, best of 5: 118 usec per loop
```

With this PR:

```
>python -m timeit -s "from pathlib import Path; p = Path(*['foo' for _ in range(1_000)])" "p.parents[100]"
Running Debug|x64 interpreter...
5 loops, best of 5: 69.4 msec per loop
```

(These numbers are on a debug build, so take them with a pinch of salt, but I think it's highly likely a non-debug build would also show a significant regression on this benchmark.)

The benchmark above is, as I say, quite artificial, but I think it's reasonable to be concerned about this. `Path.parents` must remain a property (to maintain backwards compatibility), but you generally shouldn't have properties that do a significant amount of work and could potentially become performance bottlenecks -- those should be methods rather than properties, to make it explicit that "work is being done here". (Generally speaking, nobody "expects" the "`.` operator" to do a significant amount of work.) Also, the current code does seem to work quite hard to only create new path objects when absolutely necessary, which makes me worry that this could be quite performance-sensitive.

Reply:
I've restored the `_PathParents` class. I think the comment about reference cycles is a mistake or a leftover from an early implementation.

Reply:
Yeah -- there doesn't appear to be any kind of reference cycle here, since the `.parents` property returns a new `_PathParents` object each time! So I think we're good:

https://github.com/python/cpython/blob/ae4845441281e9b6cf2623388df28587afa239c2/Lib/pathlib.py#L646-L649

Interestingly, one possible optimisation here _could_ be to make `.parents` a cached property. But that might not gain us very much (it looks to me like `_PathParents` objects should be pretty cheap to construct? But I haven't measured it), and if we did that, we really _would_ have to worry about a reference cycle.

Maybe it would be worth, though, adding a comment to the `.parents` property warning us about the possibility of a reference cycle if we were to cache the result of the property?